canonical/kubeflow-dashboard-operator
# feat: dynamic sidebar content

[Jira](https://warthogs.atlassian.net/browse/KF-567?atlOrigin=eyJpIjoiNjAxNDkzYzM1MDk1NGY1Yzk1ZTNhMTg2MTI5ZDMxMWUiLCJwIjoiaiJ9) or this #8

Follow up on this #37 (sidecar rewrite)

**Prerequisites** [katib PR](https://github.com/canonical/katib-operators/pull/8) and [tensorboards PR](https://github.com/canonical/kubeflow-tensorboards-operator/pull/16) and [jupiter-ui PR](https://github.com/canonical/notebook-operators/pull/48) and [KFP-operator PR](https://github.com/canonical/kfp-operators/pull/99)
**Use case** we want to dynamically change the content of the UI sidebar based on available application. 
**Limitation**: only for tensorboards and katib 

**Solution**: I created a `sidebar` juju relation for dashboard katib and tensorboard apps. every time a member joins the relation it sends the sibar link specification to dashboard, which adjusts the configmap. The sidebar change is instant. The opposite behavior happens on relation break.

### Setup
**Quick Dev test (excluding istio)**: run the integration tests and keep the models (check debug logs):
```
tox -e integration -- --model kubeflow --keep-models -vv -s
juju debug-log --replay --include unit-kubeflow-dashboard-0
```

**Full bundle test (including istio)**: 
- run the whole kubeflow localy `juju deploy kubeflow-lite --trust` (dont forgtet [dex configuration steps](https://charmed-kubeflow.io/docs/install))
- replace with current dashboard `juju remove-application kubeflow-dashboard`
- build current charm `charmcraft pack`
- deploy `juju deploy ./*.charm --resource oci-image=$(yq eval '.resources.oci-image.upstream-source' metadata.yaml) --trust`
- dont forget relations `juju relate kubeflow-profiles kubeflow-dashboard; juju relate istio-pilot:ingress kubeflow-dashboard:ingress`
- check the dashboard in browser

**Manual test**
- Make sure versions of `katib-ui` and `tensorboards-web-app` have the versions mentioned in **prerequisites** are are deployed in cluster
- Relate apps `juju relate tensorboards-web-app:sidebar kubeflow-dashboard:sidebar; juju relate katib-ui:sidebar kubeflow-dashboard:sidebar`
- Check the dashboard ui for sibar change or examin the configmap `k get cm centraldashboard-config -o yaml` -> data/links/menuLinks is the place for sidebar content

**Relation data**
Relation data consists of dictionary (placed in the `config` of the relation data). This dictionary can have as much key value pairs as you we want. Required pairs are:

- `type`: set to `item`
- `link`: relative path where the link will be pointing on click
- `text`: text of the tab in the sidebar (starting with capital is recommended)
- `icon`: icon which should be presented in the sidebar
- `position`: this one is optional as it specifies the position in the sidebar (lower the higher), if not specified 100 will be used 

**Bndle changes**
Make sure to merge the above mentioned PRs. Then use following changes (please fill the placeholders with correct charm versions).

```
bundle: kubernetes
name: kubeflow
applications:
   kfp-ui:
     charm: kfp-ui
     channel: <VERSION>/edge
     scale: 1
     _github_repo_name: kfp-operators
  katib-ui:
    charm: katib-ui
    channel: <VERSION>/edge
    scale: 1
    _github_repo_name: katib-operators
  tensorboards-web-app:
    charm: tensorboards-web-app
    channel: <VERSION>/edge
    scale: 1
    _github_repo_name: kubeflow-tensorboards-operator
  jupyter-ui:
    charm: jupyter-ui
    channel: <VERSION>/edge
    scale: 1
    _github_repo_name: notebook-operators
  kubeflow-volumes:
    charm: kubeflow-volumes
    channel: <VERSION>/edge
    scale: 1
    _github_repo_name: kubeflow-volumes-operator
relations:
 - [tensorboards-web-app:sidebar, kubeflow-dashboard:sidebar;]
 - [katib-ui:sidebar, kubeflow-dashboard:sidebar;]
 - [jupyter-ui:sidebar, kubeflow-dashboard:sidebar]
 - [kfp-ui:sidebar, kubeflow-dashboard:sidebar]
 - [kubeflow-volumes:sidebar, kubeflow-dashboard:sidebar]
```
